### ISSUE TITLE
pachctl update

### ISSUE THREAD

**Issue by sjezewski** [pachyderm/pachyderm#2476]
## Background

We often have users upgrading patch versions as bug fixes, etc come out. Especially if they've deployed on a cloud provider, this is error prone for users. 

A simple `pachctl undeploy` and `pachctl deploy` is insufficient.

a) It's bad UX. We require that they use all the same flags for deployment (e.g. `pachctl deploy amazon ...`). We ask that users keep the output of their original deploy command, but if they havent ... this sucks. And even if they have ... 

b) In some cases it can break their cluster. If they've modified any part of their deployment at all (e.g. some users emit `pachctl deploy amazon --dry-run ...` to a file, modify it from there (for RBAC, or other things) and deploy it via `kubectl create -f pachyderm.json`) then they'll lose any custom settings when upgrading. This is the source of a lot of user error. And can be frustrating ... because they're re-do'ing fixes or debugging scaling bottlenecks all over again.

## Goal

1) As a user, I want to upgrade patch versions (e.g. 1.6.3->1.6.4) of pachctl easily

2) As a user, I want to dump the k8s manifests of my pachyderm cluster

Even for users that follow our recommendations and save the pachyderm json manifest in a versioned repo, its very easy to make modifications and forget to update that file. This way ... you get a 'ground truth' version of what actually is there. So we make it easy for them to update said manifest in their source control.

## Iteration 0

We add two new commands:

1) `pachctl dump-cluster` which:

- queries all of the right k8s objects (e.g. `deploy/pachd`, `deploy/etcd`, `deploy/dash`, but not any worker pods) and emits a JSON file that should be identical to what you would see from a fresh `pachctl deploy --dry-run ...` command
- However, it includes any and all changes a user made to the cluster, and reflects the current state of the world

This will satisfy user story # 2. 

2) `pachctl upgrade`, which will:

- store the result of a call to `dump-cluster`
- run pachctl undeploy
- fetches the latest version of pachctl
- replaces all instances of the pachyderm version number in the stored manifest (pachd image, worker image env var, etc)
- runs kubectl create w that new manifest
- cycles any worker pods (so they get recreated w the new images)

This will satisfy user story # 1

## Batteries Not Included

- This iteration will NOT handle migrations ... users will still need to run those manually

## Future Iterations

- Support migrations
- Support minor version upgrades, e.g. 1.6.5->1.7.0

**Comment by sjezewski** [comment]
This would also come in handy for users doing more advanced deployments, e.g. using k8s namespaces: 

https://github.com/pachyderm/pachyderm/issues/2479

If we had the `pachctl dump-cluster` feature ... we could easily see how this user deployed their cluster (within a namespace). And the `pachctl update` command would modify that in place to work for their deployment.

**Comment by jdoliner** [comment]
The idea of using k8s state as a source of truth seems awful to me. Most of our major releases change our k8s assets in some non-trivial (not a version bump) way. How on earth is this going to work for those? I don't see any way it can unless we find some way to "merge" our changes with the users changes... and then we'd have to deal with potential merge conflicts... the UX on all these failure cases is really going to be awful.

I also don't agree with the basic premise of this issue. I don't really see the `pachctl undeploy` `pachctl deploy` flow as deficient except in cases where you're manually modifying the manifest. Having to store a set of command line with arguments in version control so you can deploy the same way again doesn't seem unreasonable at all. What if the cluster gets destroyed by an outage or an intern or something? Users should enable themselves to deploy the same cluster again and the currently running cluster is not a good place to store that information. That information is code that should be in version control, not stored in live servers. Cases where a user has manually modified the cluster are, admittedly more difficult but I think a) those cases are relatively rare, b) we can't do anything about them. There's no consistent way we can reapply those changes. If users want to make their manual modifications to k8s manifests automated then they're going to have to automate them themselves.

**Comment by sjezewski** [comment]
Interesting point. Why is that awful? It seems like that's what k8s does for their components. It sounds like your objection is to the automation here (the undeploy / redeploy), not actually using k8s as a source of truth.

And to that point - I agree - it's not automatable. You're going to need a human to do any sort of 'merge conflict' on our k8s deployment. So scratch that first user story.

However, I do think it would be helpful to be able to 'read' the state of the cluster. (User story 2). This enables me as a user to do two things:

a) debugability

This would give a human the tool they need to make doing this sort of 'merge conflict' resolution possible. This sort of semantic is analogous to the tooling that `kubectl get ...` or `kops edit` enables. It's so easy to forget what you modify in the course of development/testing. Personally I've found that tooling invaluable - for debugging and for development. And I agree ... the best practice should always be storing this info in version control. But that's not how folks develop. I don't check in a k8s manifest every time I'm debugging a cluster or testing a deployment. Especially while standing something up for the first time.

This will make our user's lives easier. And this will make it much easier to support our users when they come to us w an issue. I could ask for the output of `pachctl dump-cluster` to see exactly what's going on. There are a bunch of use cases that are not custom deployments that this would help debug (e.g. debugging that they setup their obj store secret correctly)

b) manage custom deployments

While this use case (user's manually modifying the manifest) has been rare in the past ... we've seen this come up more recently. And since we'll never want to support all of the k8s api surface via `pachctl`, I don't see this 'manual modifications' use case going away or lessening anytime soon. 

Some recent examples include:

- A user adding custom RBAC config
- A user adding custom namespacing

Again - an alternative would be to support each of these things through our API. But I doubt we want to do that. K8s is great at providing options for deployments. There are a bunch of k8s tools that promote good practices for true production grade clusters. That's not our core value proposition. And we don't want a bunch of flag/options inflating our deployment commands.

**Comment by jdoliner** [comment]
> Interesting point. Why is that awful? It seems like that's what k8s does for their components.

I think ultimately this comes down to controlling the data format. K8s assets are a format that k8s controls so it makes for them to use those as their source of truth, when the schema changes they understand the changes and can somewhat easily write migration code. That being said I don't think they even really do this consistently. We commonly have to change our assets to support new versions of k8s because they've changed the format in non-backward-compatible ways.

In terms of `dump-cluster` it sounds like most of the functionality you're describing is covered by `kubectl get all -o json -l suite=pachyderm` (ignoring the fact that `all` doesn't actually mean `all`) I suppose we could add a comman in `pachctl` which just sends the right command to k8s to dump stuff. This doesn't seem a lot better to me than just having them do the k8s command themselves. We might save them some typing, but I think having them type the actual command helps them understand the underlying k8s system a bit better which is helpful. Especially since this is designed for people who have already manually edited their k8s manifest it's likely they'll already know how to do this. I also want to be clear that this doesn't help with upgrading at all, you'll still need to reapply your manifest changes to a new manifest generated by the new version of `pachctl`.

**Comment by sjezewski** [comment]
Yea, either way -- either a user makes the call for k8s to dump the right stuff ... or we have a helper that does it. 

It seems reasonable to me to have a helper. Or at least, for us to apply a label to the Right Things so that a very simple `kubectl` command gets the info we need. The command you list would also dump the worker pod specs, which while that might be interesting, was not the goal here. The goal here would be to get the equivalent of the k8s objects you'd see in the manifest output by `pachctl deploy --dry-run ...`, so we'd see the deployments / RCs, but not the worker pod specs. The worker pod specs should be wholly determined by our pipeline spec format. [Which is why we'd want a dump command for our pipelines](https://github.com/pachyderm/pachyderm/issues/2477). 

(Though I think I heard talk of some sort of `--dry-run` support for debugging worker pod specs? I'm not sure what the use case was there. Perhaps users need to apply custom k8s options on those specs as well, e.g. namespacing. I'm not sure.)

I don't think it doesn't help upgrading at all. I think it does help. Yes, in these cases, there will always be manual editing. But the manual case is what we're trying to make easier. If a user has made extensive changes, it might be easier to start w the manifest from their dump, and add the new objects from the output of `pachctl`. In other cases, it will work as you describe. They'll have to look through their dump to see what's different, and add that again to the new manifest. But without seeing the state of what they have, they can omit configurations that they need. Without the ability to see this state, they can easily shoot themselves in the foot. 

To that end, as a user, if the objects weren't reported in the same order (`pachctl deploy --dry-run` vs `kubectl get all ... -o json`), I'd additionally want to run them through some normalization (lets say ... sort the objects) so that I could easily diff those two results to see if there were any adjustments I'd need to make manually. 

So AFAICT, you could get close to this result today w some `kubectl` commands and bash-fu. But to me this is a UX thing. It seems pretty reasonable to make it easy for users to do this. The more often they can do this sort of things themselves, the further they will get w the product.

So to be clear, at this point I'm proposing:

- a pachctl command that issues queries via `kubectl` for the right k8s objects in the right order,
- and outputs the result as JSON

Otherwise, we're falling back on recommending a convention to users. (Using a kubectl command, etc). But that hasn't been working for us so far (telling users to version control their manifests). I get that we want users to be better at k8s. This is tricky. Ultimately we don't want to be in the business of training data engineers, we want folks to use our hosted version. Till then, we just want folks to use our tech.

**Comment by jdoliner** [comment]
This issue is mostly implemented. We wound up calling it `extract` instead of `dump`. The only thing from this issue that's missing is that users need to replicate their `deploy` arguments in order to upgrade.
